<a href="https://colab.research.google.com/github/praffs18/Telecom-Churn-Analysis/blob/main/Telecom_Churn_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**    - Telecom Churn Analysis



##### **Project Type**    - EDA
##### **Contribution**    - Team
##### **Team Member 1 -** Prafull Korde
##### **Team Member 2 -** 


# **Project Summary -**

Write the summary here within 500-600 words.

# **GitHub Link -**

Provide your GitHub Link here.

# **Problem Statement**


Telecom churn analysis is the process of identifying customers who are likely to cancel their service or switch to a different service provider. This is an important problem for telecom companies, as churn can have a significant impact on their revenue and profitability.

Orange S.A., formerly France Télécom S.A., is a French multinational telecommunications corporation. The Orange Telecom's Churn Dataset, consists of cleaned customer activity data (features), along with a churn label specifying whether a customer canceled the subscription.
Explore and analyze the data to discover key factors responsible for customer churn and come up with ways/recommendations to ensure customer retention.

#### **Define Your Business Objective?**

*   Identifying the key cause of the customer churn
*   Provide steps to retain the valuable customer

# **General Guidelines** : -  

1.   Well-structured, formatted, and commented code is required. 
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits. 
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 20 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule. 

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np

# Import Visualization Libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# Import warnings
import warnings
warnings.filterwarnings('ignore')

### Dataset Loading

In [2]:
# Load Dataset
url = 'https://raw.githubusercontent.com/praffs18/Telecom-Churn-Analysis/main/Telecom%20Churn.csv'
data = pd.read_csv(url)

### Dataset First View

In [3]:
# Dataset First Look
data.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [4]:
data.tail()

State  Account length  Area code International plan Voice mail plan  \
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
3328                       2  False  
3329                       3  False  
3330                       2  False  
3331                       2  False  
3332                       0  False

### Dataset Rows & Columns count

In [5]:
# Dataset Rows & Columns count
print("Number of rows are: ",data.shape[0])
print("Number of columns are: ",data.shape[1])

Number of rows are:  3333
Number of columns are:  20


### Dataset Information

In [6]:
# Dataset Info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

#### Duplicate Values

In [7]:
# Dataset Duplicate Value Count
data.duplicated().sum()

0

## There are no duplicate valuse present

#### Missing Values/Null Values

In [8]:
# Missing Values/Null Values Count
data.isnull().sum()

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

### What did you know about your dataset?

This dataset is from telecom industry. and we know that there is threat to the company when customer churn happaens. So we have to analyze the customer churn and insights of it. and study the product or services which are causing the churn.

This dataset is having 3333 no. of rows and 20 feature columns.
and there are no missing values and duplicate values in dataset.


## ***2. Understanding Your Variables***

In [9]:
# Dataset Columns
data.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [10]:
# Dataset Describe
data.describe(include= 'all').round(2)

State  Account length  Area code International plan Voice mail plan  \
count   3333         3333.00    3333.00               3333            3333   
unique    51             NaN        NaN                  2               2   
top       WV             NaN        NaN                 No              No   
freq     106             NaN        NaN               3010            2411   
mean     NaN          101.06     437.18                NaN             NaN   
std      NaN           39.82      42.37                NaN             NaN   
min      NaN            1.00     408.00                NaN             NaN   
25%      NaN           74.00     408.00                NaN             NaN   
50%      NaN          101.00     415.00                NaN             NaN   
75%      NaN          127.00     510.00                NaN             NaN   
max      NaN          243.00     510.00                NaN             NaN   

        Number vmail messages  Total day minutes  Total day calls  \
count                 3333.00            3333.00          3333.00   
unique                    NaN                NaN              NaN   
top                       NaN                NaN              NaN   
freq                      NaN                NaN              NaN   
mean                     8.10             179.78           100.44   
std                     13.69              54.47            20.07   
min                      0.00               0.00             0.00   
25%                      0.00             143.70            87.00   
50%                      0.00             179.40           101.00   
75%                     20.00             216.40           114.00   
max                     51.00             350.80           165.00   

        Total day charge  Total eve minutes  Total eve calls  \
count            3333.00            3333.00          3333.00   
unique               NaN                NaN              NaN   
top                  NaN                NaN              NaN   
freq                 NaN                NaN              NaN   
mean               30.56             200.98           100.11   
std                 9.26              50.71            19.92   
min                 0.00               0.00             0.00   
25%                24.43             166.60            87.00   
50%                30.50             201.40           100.00   
75%                36.79             235.30           114.00   
max                59.64             363.70           170.00   

        Total eve charge  Total night minutes  Total night calls  \
count            3333.00              3333.00            3333.00   
unique               NaN                  NaN                NaN   
top                  NaN                  NaN                NaN   
freq                 NaN                  NaN                NaN   
mean               17.08               200.87             100.11   
std                 4.31                50.57              19.57   
min                 0.00                23.20              33.00   
25%                14.16               167.00              87.00   
50%                17.12               201.20             100.00   
75%                20.00               235.30             113.00   
max                30.91               395.00             175.00   

        Total night charge  Total intl minutes  Total intl calls  \
count              3333.00             3333.00           3333.00   
unique                 NaN                 NaN               NaN   
top                    NaN                 NaN               NaN   
freq                   NaN                 NaN               NaN   
mean                  9.04               10.24              4.48   
std                   2.28                2.79              2.46   
min                   1.04                0.00              0.00   
25%                   7.52                8.50              3.00   
50%                   9.05               10

### Variables Description 

State :categorica for the 50 states and capital DC

Account Length :how long account has been active

Area Code :Code Number of Area having some States included in each area code

lntl Plan :Internat ional plan activated ( yes, no )

VMail Plan : ice Mail plan activated ( yes ,no )

VMail Message :No.of voice mail messages

Day Mins :Total day minutes used

Day calls :Total day calls made

Day Charge :Total day charge

Eve Mins :Total evening minutes

Eve Calls :Total evening calls

Eve Charge :Total evening charge

Night Mins :Total night minutes

Night Calls :Total night calls

Night Charge :Total night charge

Intl Mins :Total International minutes used

Intl Calls :Total International calls made

Intl Charge :Total International charge

CustServ calls :Number of customer service caUs made

Churn :Customer churn (Target Variable True=1, False=0)Answer Here

### Check Unique Values for each variable.

In [11]:
# Check Unique Values for each variable.
for i in data.columns.tolist():
  print("No. of unique values in",i,"is",data[i].nunique())

No. of unique values in State is 51
No. of unique values in Account length is 212
No. of unique values in Area code is 3
No. of unique values in International plan is 2
No. of unique values in Voice mail plan is 2
No. of unique values in Number vmail messages is 46
No. of unique values in Total day minutes is 1667
No. of unique values in Total day calls is 119
No. of unique values in Total day charge is 1667
No. of unique values in Total eve minutes is 1611
No. of unique values in Total eve calls is 123
No. of unique values in Total eve charge is 1440
No. of unique values in Total night minutes is 1591
No. of unique values in Total night calls is 120
No. of unique values in Total night charge is 933
No. of unique values in Total intl minutes is 162
No. of unique values in Total intl calls is 21
No. of unique values in Total intl charge is 162
No. of unique values in Customer service calls is 10
No. of unique values in Churn is 2


## 3. ***Data Wrangling***

### Data Wrangling Code

In [12]:
# Write your code to make your dataset analysis ready.
# Before doing any data wrangling lets create copy 
data_2 = data.copy()

In [13]:
# Check for how many customer leave the company
churn = pd.DataFrame(data_2['Churn'].value_counts())
churn['Churn Rate'] = round(churn*100/len(data_2),2)
churn

Churn  Churn Rate
False   2850       85.51
True     483       14.49

In [14]:
# Lets check churn rate area code wise
churn_area = pd.DataFrame(data.groupby('Area code')['Churn'].value_counts())
churn_area['churn rate'] = round(churn_area*100/len(data_2),2)
churn_area

Churn  churn rate
Area code Churn                   
408       False    716       21.48
          True     122        3.66
415       False   1419       42.57
          True     236        7.08
510       False    715       21.45
          True     125        3.75

In [15]:
#Lets check the statistical value(mean, median) of all feature of area code wise
def get_mean_median(data,area_code):
  ''' This fuction gives mean and median of all features'''
  try:
    return pd.concat([data_2[data_2['Area code']==area_code].describe().round(2).iloc[1],
          data_2[data_2['Area code']==area_code].describe().round(2).iloc[5]], axis = 1).rename(columns={'50%':'Median'})
  except:
    print("Invalid Area code")

In [16]:
# For areacode 408
get_mean_median(data_2,408)

mean  Median
Account length          101.88  102.00
Area code               408.00  408.00
Number vmail messages     7.67    0.00
Total day minutes       177.18  176.35
Total day calls         100.50  100.00
Total day charge         30.12   29.98
Total eve minutes       201.28  202.50
Total eve calls          99.79   99.00
Total eve charge         17.11   17.21
Total night minutes     199.21  199.55
Total night calls        99.04  100.00
Total night charge        8.96    8.98
Total intl minutes       10.13   10.10
Total intl calls          4.43    4.00
Total intl charge         2.73    2.73
Customer service calls    1.52    1.00

In [17]:
#For areacode = 415
get_mean_median(data_2,415)

mean  Median
Account length          101.07  100.00
Area code               415.00  415.00
Number vmail messages     8.36    0.00
Total day minutes       181.59  180.70
Total day calls         100.58  101.00
Total day charge         30.87   30.72
Total eve minutes       200.65  200.70
Total eve calls         100.50  101.00
Total eve charge         17.06   17.06
Total night minutes     202.04  203.00
Total night calls       100.40  101.00
Total night charge        9.09    9.14
Total intl minutes       10.34   10.50
Total intl calls          4.56    4.00
Total intl charge         2.79    2.84
Customer service calls    1.55    1.00

In [18]:
#For areacode = 510
get_mean_median(data_2,510)

mean  Median
Account length          100.25  100.00
Area code               510.00  510.00
Number vmail messages     8.02    0.00
Total day minutes       178.79  179.45
Total day calls         100.10  100.00
Total day charge         30.39   30.51
Total eve minutes       201.32  201.95
Total eve calls          99.67  100.50
Total eve charge         17.11   17.16
Total night minutes     200.23  197.95
Total night calls       100.60  100.00
Total night charge        9.01    8.91
Total intl minutes       10.14   10.20
Total intl calls          4.37    4.00
Total intl charge         2.74    2.75
Customer service calls    1.62    1.00

In [19]:
# Churn count by statewise
churn_state = pd.DataFrame(data_2.groupby('State')['Churn'].value_counts())
churn_state

Churn
State Churn       
AK    False     49
      True       3
AL    False     72
      True       8
AR    False     44
...            ...
WI    True       7
WV    False     96
      True      10
WY    False     68
      True       9

[102 rows x 1 columns]

In [20]:
# Lets check churn on the basis of International plan
churn_int = pd.DataFrame(data_2.groupby('International plan')['Churn'].value_counts())
churn_int['churn rate'] = round(churn_int*100/len(data_2),2)
churn_int

Churn  churn rate
International plan Churn                   
No                 False   2664       79.93
                   True     346       10.38
Yes                False    186        5.58
                   True     137        4.11

In [21]:
#Lets check churn on the basis of Voice mail plan
churn_vmail = pd.DataFrame(data_2.groupby('Voice mail plan')['Churn'].value_counts())
churn_vmail['churn rate'] = round(churn_vmail*100/len(data_2),2)
churn_vmail

Churn  churn rate
Voice mail plan Churn                   
No              False   2008       60.25
                True     403       12.09
Yes             False    842       25.26
                True      80        2.40

### What all manipulations have you done and insights you found?

Answer Here.

## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

#### Chart - 1

In [22]:
# Chart - 1 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 2

In [23]:
# Chart - 2 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 3

In [24]:
# Chart - 3 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 4

In [25]:
# Chart - 4 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 5

In [26]:
# Chart - 5 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 6

In [27]:
# Chart - 6 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 7

In [28]:
# Chart - 7 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 8

In [29]:
# Chart - 8 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 9

In [30]:
# Chart - 9 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 10

In [31]:
# Chart - 10 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 11

In [32]:
# Chart - 11 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 12

In [33]:
# Chart - 12 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 13

In [34]:
# Chart - 13 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 14 - Correlation Heatmap

In [35]:
# Correlation Heatmap visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

#### Chart - 15 - Pair Plot 

In [36]:
# Pair Plot visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

## **5. Solution to Business Objective**

#### What do you suggest the client to achieve Business Objective ? 
Explain Briefly.

Answer Here.

# **Conclusion**

Write the conclusion here.

### ***Hurrah! You have successfully completed your EDA Capstone Project !!!***